README
NAME
    Data::GUID::Any - Generic interface for GUID creation

VERSION
    This documentation describes version 0.002.

SYNOPSIS
         use Data::GUID::Any 'guid_as_string';
     
     my $guid = guid_as_string();

DESCRIPTION
    This module is a generic wrapper around various ways of obtaining
    Globally Unique ID's (GUID's). It will use any of the following, listed
    from most preferred to least preferred:

    *   Data::GUID

    *   Data::UUID

    *   Data::UUID::LibUUID

    *   UUID

    *   Win32 (using GuidGen())

    *   APR::UUID (random)

    *   UUID::Random (random)

    *   uuid (external program)

    If none are available when Data::GUID::Any is installed, it will add
    Data::GUID as a prerequisite.

USAGE
  guid_as_string()
         my $guid = guid_as_string();

    Returns a guid in string format with upper-case hex characters:

       FA2D5B34-23DB-11DE-B548-0018F34EC37C

    Except for modules that only produce random GUID's, these are 'version
    1' GUID's.

BUGS
    Please report any bugs or feature requests using the CPAN Request
    Tracker web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=Data-GUID-Any>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   RFC 4122 <http://tools.ietf.org/html/rfc4122>

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2009 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Data-GUID-Any

{{$NEXT}}

0.005     2018-04-22 15:23:46+02:00 Europe/Oslo

    [FIXED]

    - Fixed tests for Perls without . in @INC

0.004     2012-12-03 11:59:00 America/New_York

    - Documented that uppercase strings are not compliant with RFC 4122

    - Added $Data::GUID::Any::UC variable that can be set false to get
      lowercase strings instead

0.003     2012-04-18 12:59:58 America/New_York
    
    - The ordering of GUID providers  has changed and there are now
      separate functions for generic GUID (any supported type), v1 and v4
      type GUIDs.
    
    - Data::UUID::MT is now preferred, since it offers good speed and the
      best control.

    - The uuid binary is now preferred to pure-perl UUID::Tiny.

    - UUID::Generator::PurePerl has been dropped as the author has listed
      it as deprecated.
    
    - distribution converted to Dist::Zilla style

0.002 Fri Apr 10 10:19:55 EDT 2009
    - added UUID::Generator::PurePerl
    - cleaned up some internals

0.001 Tue Apr  7 23:33:51 EDT 2009
    - initial release


t/Util.pm
package Util;
use strict;
use warnings;
use Exporter;
our @ISA = qw/Exporter/;
our @EXPORT = qw/looks_like_uc_guid looks_like_lc_guid/;

my $uchex = "A-Z0-9";
my $lchex = "a-z0-9";

sub looks_like_uc_guid {
  my $guid = shift;
  return $guid =~ /[$uchex]{8}-[$uchex]{4}-[$uchex]{4}-[$uchex]{4}-[$uchex]{12}/;
}

sub looks_like_lc_guid {
  my $guid = shift;
  return $guid =~ /[$lchex]{8}-[$lchex]{4}-[$lchex]{4}-[$lchex]{4}-[$lchex]{12}/;
}

1;


Build.PL
# Copyright (c) 2009 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use 5.006;
use strict;
use warnings;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";

my @maybe_reqs;

unless ( eval { require 'lib/Data/GUID/Any.pm'; 1 } ) {
  push @maybe_reqs, 'Data::GUID', 0;
}

my $build = $class->new( 
  module_name => 'Data::GUID::Any',
  dist_author => 'David Golden <dagolden@cpan.org>',
  license => 'apache',
  create_readme => 1,
  requires => {
    'perl' => 5.006,
    'base' => 0, 
    'Exporter' => 0,
    'File::Spec' => 3.12, 
    @maybe_reqs,
  },
  build_requires => {
    'File::Find' => 0,
    'Test::More' => 0.62, 
  },
  meta_add            => { 
    configure_requires => { 
      'Module::Build' => 0.30 
    },
    no_index => { 
      directory => [ qw/ examples inc t xt /],
    },
    resources => {
      repository => "http://github.com/dagolden/data-guid-any/", 
      bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-GUID-Any",
    },
  },
);
$build->create_build_script;



Makefile.PL
# inc::MakeMaker <self>
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.41.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Generic interface for GUID/UUID creation",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::CBuilder" => "0.26",
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Data-GUID-Any",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Data::GUID::Any",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "IPC::Cmd" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.92",
    "lib" => 0
  },
  "VERSION" => "0.006",
  "test" => {
    "TESTS" => "t/*.t"
  }
);
;

$WriteMakefileArgs{PREREQ_PM} ||= {};

eval {
  local @INC = ('lib', @INC);
  require Data::GUID::Any;
  Data::GUID::Any::v1_guid_as_string(); # dies if no provider
  Data::GUID::Any::v4_guid_as_string(); # dies if no provider
  1;
} or do {
  require ExtUtils::CBuilder;
  if ( ExtUtils::CBuilder->new->have_compiler ) {
    $WriteMakefileArgs{PREREQ_PM}{'Data::UUID::MT'} = '0';
  }
  else {
    $WriteMakefileArgs{PREREQ_PM}{'UUID::Tiny'} = '0';
  }
};

# Hey, CPAN Testers, go ahead and test with extra prereqs
# that don't have external library dependencies
if ( $ENV{AUTOMATED_TESTING} ) {
  require ExtUtils::CBuilder;
  $WriteMakefileArgs{BUILD_REQUIRES}{'Data::UUID::MT'} = '0'
    if ExtUtils::CBuilder->new->have_compiler;
  $WriteMakefileArgs{BUILD_REQUIRES}{'UUID::Tiny'} = '0';
}
 

my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IPC::Cmd" => 0,
  "Test::More" => "0.92",
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


inc/MakeMaker.pm
package inc::MakeMaker;
use Moose;
extends 'Dist::Zilla::Plugin::MakeMaker::Awesome';
 
use namespace::autoclean;
 
override _build_WriteMakefile_dump => sub {
  my ($self) = @_;
 
  my $str = super;
 
  $str .= ";\n\n";
 
  $str .= <<'END_NONSENSE';
$WriteMakefileArgs{PREREQ_PM} ||= {};

use ExtUtils::CBuilder;
my $have_compiler = ExtUtils::CBuilder->new->have_compiler;
 
my $have_providers = eval {
  use lib 'lib';
  require Data::GUID::Any;
  Data::GUID::Any::v1_guid_as_string(); # dies if no provider
  Data::GUID::Any::v4_guid_as_string(); # dies if no provider
  1;
};

if ( $have_providers ) {
  # nothing to add
}
elsif ( $have_compiler ) {
  $WriteMakefileArgs{PREREQ_PM}{'Data::UUID::MT'} = '0';
}
else {
  $WriteMakefileArgs{PREREQ_PM}{'UUID::Tiny'} = '0';
}
 
END_NONSENSE
 
  return $str;
};
 
1;


lib/Data/GUID/Any.pm
use 5.006;
use strict;
use warnings;
package Data::GUID::Any;
# ABSTRACT: Generic interface for GUID/UUID creation

our $VERSION = '0.006';

use IPC::Cmd;
use Exporter;
our @ISA = qw/Exporter/;
our @EXPORT_OK = qw/ guid_as_string v1_guid_as_string v4_guid_as_string/;

our ($Using_vX, $Using_v1, $Using_v4) = ("") x 3;
our $UC = 1;

#--------------------------------------------------------------------------#

my $hex = "a-z0-9";

# case insensitive, since used to check if generators are functioning
sub _looks_like_guid {
  my $guid = shift;
  return $guid =~ /[$hex]{8}-[$hex]{4}-[$hex]{4}-[$hex]{4}-[$hex]{12}/i;
}

#--------------------------------------------------------------------------#

sub _xc {
  return $UC ? uc($_[0]) : lc($_[0]);
}

#--------------------------------------------------------------------------#

# state variables for generator closures
my ($dumt_v1, $dumt_v4, $uuid_v1, $uuid_v4) = (undef) x 4; # reset if reloaded

my %generators = (
  # v1 or v4
  'Data::UUID::MT' => {
    type => 'module',
    v1 => sub {
      $dumt_v1 ||= Data::UUID::MT->new(version => 1);
      return _xc( $dumt_v1->create_string );
    },
    v4 => sub {
      $dumt_v4 ||= Data::UUID::MT->new(version => 4);
      return _xc( $dumt_v4->create_string );
    },
  },
  'Data::UUID::LibUUID' => {
    type => 'module',
    v1 => sub { return _xc( Data::UUID::LibUUID::new_uuid_string(2) ) },
    v4 => sub { return _xc( Data::UUID::LibUUID::new_uuid_string(4) ) },
    vX => sub { return _xc( Data::UUID::LibUUID::new_uuid_string() ) },
  },
  'UUID::Tiny' => {
    type => 'module',
    v1 => sub { return _xc( UUID::Tiny::create_UUID_as_string(UUID::Tiny::UUID_V1()) ) },
    v4 => sub { return _xc( UUID::Tiny::create_UUID_as_string(UUID::Tiny::UUID_V4()) ) },
  },
  'uuid' => {
    type => 'binary',
    v1 => sub {
      $uuid_v1 ||= IPC::Cmd::can_run('uuid');
      chomp( my $guid = qx/$uuid_v1 -v1/ ); return _xc( $guid );
    },
    v4 => sub {
      $uuid_v4 ||= IPC::Cmd::can_run('uuid');
      chomp( my $guid = qx/$uuid_v4 -v4/ ); return _xc( $guid );
    },
  },
  # v1 only
  'Data::GUID' => {
    type => 'module',
    v1 => sub { return _xc( Data::GUID->new->as_string ) },
  },
  'Data::UUID' => {
    type => 'module',
    v1 => sub { return _xc( Data::UUID->new->create_str ) },
  },
  # system dependent or custom
  'UUID' => {
    type => 'module',
    vX => sub { my ($u,$s); UUID::generate($u); UUID::unparse($u, $s); return _xc( $s ) },
  },
  'Win32' => {
    type => 'module',
    vX => sub { my $guid = Win32::GuidGen(); return _xc( substr($guid,1,-1) ) },
  },
  'APR::UUID' => {
    type => 'module',
    vX => sub { return _xc( APR::UUID->new->format ) },
  },
);

our $NO_BINARY; # for testing
sub _is_available {
  my ($name) = @_;
  if ( $generators{$name}{type} eq 'binary' ) {
    return $NO_BINARY ? undef : IPC::Cmd::can_run($name);
  }
  else {
    return eval "require $name";
  }
}

sub _best_generator {
  my ($list) = @_;
  for my $option ( @$list ) {
    my ($name, $version) = @$option;
    next unless my $g = $generators{$name};
    next unless _is_available($name);
    return ($name, $g->{$version})
      if $g->{$version} && _looks_like_guid( $g->{$version}->() );
  }
  return;
}

#--------------------------------------------------------------------------#

my %sets = (
  any => [
    ['Data::UUID::MT'       => 'v4'],
    ['Data::GUID'           => 'v1'],
    ['Data::UUID'           => 'v1'],
    ['Data::UUID::LibUUID'  => 'vX'],
    ['UUID'                 => 'vX'],
    ['Win32'                => 'vX'],
    ['uuid'                 => 'v1'],
    ['APR::UUID'            => 'vX'],
    ['UUID::Tiny'           => 'v1'],
  ],
  v1 => [
    ['Data::UUID::MT'       => 'v1'],
    ['Data::GUID'           => 'v1'],
    ['Data::UUID'           => 'v1'],
    ['Data::UUID::LibUUID'  => 'v1'],
    ['uuid'                 => 'v1'],
    ['UUID::Tiny'           => 'v1'],
  ],
  v4 => [
    ['Data::UUID::MT'       => 'v4'],
    ['Data::UUID::LibUUID'  => 'v4'],
    ['uuid'                 => 'v4'],
    ['UUID::Tiny'           => 'v4'],
  ],
);

sub _generator_set { return $sets{$_[0]} }

{
  no warnings qw/once redefine/;
  {
    my ($n, $s) = _best_generator(_generator_set("any"));
    die "Couldn't find a GUID provider" unless $n;
    *guid_as_string = $s;
    $Using_vX = $n;
  }
  {
    my ($n, $s) = _best_generator(_generator_set("v1"));
    *v1_guid_as_string = $s || sub { die "No v1 GUID provider found\n" };
    $Using_v1 = $n || '';
  }
  {
    my ($n, $s) = _best_generator(_generator_set("v4"));
    *v4_guid_as_string = $s || sub { die "No v4 GUID provider found\n" };
    $Using_v4 = $n || '';
  }
}

1;

__END__

=head1 SYNOPSIS

    use Data::GUID::Any 'guid_as_string';

    my $guid = guid_as_string();

=head1 DESCRIPTION

This module is a generic wrapper around various ways of obtaining Globally
Unique ID's (GUID's), also known as Universally Unique Identifiers (UUID's).

On installation, if Data::GUID::Any can't detect a way of generating both
version 1 and version 4 GUID's, it will add either Data::UUID::MT or UUID::Tiny
as a prerequisite, depending on whether or not a compiler is available.

For legacy compatibility with L<Data::UUID>, guid strings are returned uppercase,
even though RFC 4122 specifies that generators should provide lower-case strings.
To force lower case results from Data::GUID::Any, set C<$Data::GUID::Any::UC>
to a false value.

  local $Data::GUID::Any::UC;
  guid_as_string(); # will be lower case

=head1 USAGE

The following functions are available for export.

=head2 guid_as_string()

    my $guid = guid_as_string();

Returns a guid in string format with upper-case hex characters:

  FA2D5B34-23DB-11DE-B548-0018F34EC37C

This is the most general subroutine that offers the least amount of control
over the result.  This routine returns whatever is the default type of GUID for
a source, which could be version 1 or version 4 (or, in the case of Win32,
something resembling a version 1, but specific to Microsoft).

It will use any of the following sources, listed from most preferred to least
preferred:

=for :list
* L<Data::UUID::MT> (v4)
* L<Data::GUID> (v1)
* L<Data::UUID> (v1)
* L<Data::UUID::LibUUID> (v4 or v1)
* L<UUID> (v4 or v1)
* L<Win32> (using GuidGen()) (similar to v1)
* uuid (external program) (v1)
* L<APR::UUID> (v4 or v1)
* L<UUID::Tiny> (v1)

At least one of them is guaranteed to exist or Data::GUID::Any will
throw an exception when loaded. This shouldn't happen if prerequisites
were correctly installed.

=head2 v1_guid_as_string()

    my $guid = v1_guid_as_string();

Returns a version 1 (timestamp+MAC/random-identifier) GUID in string format
with upper-case hex characters from one of the following sources:

=for :list
* L<Data::UUID::MT>
* L<Data::GUID>
* L<Data::UUID>
* L<Data::UUID::LibUUID>
* uuid (external program)
* L<UUID::Tiny>

If none of them are available, an exception will be thrown when this is called.
This shouldn't happen if prerequisites were correctly installed.


=head2 v4_guid_as_string()

    my $guid = v4_guid_as_string();

Returns a version 4 (random) GUID in string format with upper-case hex
characters from one of the following modules:

=for :list
* L<Data::UUID::MT>
* L<Data::UUID::LibUUID>
* uuid (external program)
* L<UUID::Tiny>

If none of them are available, an exception will be thrown when this is called.
This shouldn't happen if prerequisites were correctly installed.

=head1 SEE ALSO

=for :list
* RFC 4122 [http://tools.ietf.org/html/rfc4122]

=cut

